aptos-move/aptos-vm/src/sharded_block_executor/sharded_executor_client.rs

```diff
@@ -0,0 +1,197 @@
+        for txn in &sub_block.transactions {
+            for (_, storage_locations) in txn.cross_shard_dependencies.required_edges_iter() {
+                for storage_location in storage_locations {
+                    cross_shard_state_key.insert(storage_location.clone().into_state_key());
+                }
+            }
+        }
+        CrossShardStateView::new(self.shard_id, cross_shard_state_key, base_view)
+    }
+
+    fn execute_sub_block<S: StateView + Sync + Send>(
+        &self,
+        sub_block: SubBlock<Transaction>,
+        round: usize,
+        state_view: &S,
+        concurrency_level: usize,
+        maybe_block_gas_limit: Option<u64>,
+    ) -> Result<Vec<TransactionOutput>, VMStatus> {
+        trace!(
+            "executing sub block for shard {} and round {}",
+            self.shard_id,
+            round
+        );
+        let cross_shard_commit_sender = CrossShardCommitSender::new(
+            self.shard_id,
+            self.message_txs
+                .iter()
+                .map(|t| t.lock().unwrap().clone())
+                .collect(),
+            &sub_block,
+        );
+
+        let (callback, callback_receiver) = oneshot::channel();
+
+        let message_rxs = self.message_rx.clone();
+        let self_message_tx = Arc::new(Mutex::new(
+            self.message_txs[self.shard_id].lock().unwrap().clone(),
+        ));
+        let cross_shard_state_view =
+            Arc::new(self.create_cross_shard_state_view(state_view, &sub_block));
+        let cross_shard_state_view_clone1 = cross_shard_state_view.clone();
+        self.executor_thread_pool.scope(|s| {
+            s.spawn(move |_| {
+                if round != 0 {
+                    // If this is not the first round, start the cross-shard commit receiver.
+                    // this is a bit ugly, we can get rid of this when we have round number
+                    // information in the cross shard dependencies.
+                    CrossShardCommitReceiver::start(
+                        cross_shard_state_view_clone1,
+                        &message_rxs.lock().unwrap(),
+                    );
+                }
+            });
+            s.spawn(move |_| {
+                let ret = BlockAptosVM::execute_block(
+                    self.executor_thread_pool.clone(),
+                    BlockExecutorTransactions::Unsharded(sub_block.into_txns()),
+                    cross_shard_state_view.as_ref(),
+                    concurrency_level,
+                    maybe_block_gas_limit,
+                    Some(cross_shard_commit_sender),
+                );
+                // Send a stop command to the cross-shard commit receiver.
+                if round != 0 {
+                    self_message_tx
+                        .lock()
+                        .unwrap()
+                        .send(CrossShardMsg::StopMsg)
+                        .unwrap();
+                }
+                callback.send(ret).unwrap();
```

Comment:
does this have to be in a separate thread? I wonder by putting it in a separate thread we further lose another thread...

Reply:
You are right, this doesn't need to be in a separate thread, I can change that. But we are not loosing a thread as we are spawning two additional threads.

Reply:
@zjma - Actually I take it back, the execution needs to be in a separate thread as we are using scope() here, this blocks until all the task spawned inside is finished. If we move execution to outside of the scope(), then basically there is a deadlock.